python-lib/wordcloud_generator.py

```diff
@@ -0,0 +1,192 @@
+        elif self.subchart_column != None:
+            self.subcharts = self.group_names
+            self.languages = [self.language] * len(self.subcharts)
+        else:
+            self.languages = self.group_names
+
+        # Tokenize
+        self.docs = [
+            self.tokenizer.tokenize_list(text, language)[0] for text, language in zip(self.texts, self.languages)
+        ]
+        logging.info("Tokenization done in {:.2f} seconds.".format(time() - start))
+
+    def _count_tokens(self):
+        # Count tokens in each group
+        start = time()
+        logging.info("Initializing count")
+        self.counters = []
+        for doc in self.docs:
+            counter = Counter()
+            for token in doc:
+                counter[(token.text)] += 1  # Equivalently, token.lemma_
+            self.counters.append(counter)
+        logging.info("Count successful, aggregating counters according to chart settings")
+
+        if self.subchart_column == None:
+            # sum the values with same keys
+            self.counts = Counter()
+            for d in self.counters:
+                self.counts.update(d)
+
+            self.counts = dict(self.counts)
+        else:
+            self.counts_df = pd.DataFrame(
+                list(zip(self.subcharts, self.counters)),
+                columns=["subchart", "count"],
+            )
+            self.counts_df = self.counts_df.groupby(by=["subchart"]).agg({"count": "sum"})
+            # remove subcharts emptied by filter
+            self.counts_df = self.counts_df.loc[self.counts_df["count"] != {}, :]
+
+        logging.info("Counter aggregation successful, counting done in {:.2f} seconds.".format(time() - start))
+
+    def _generate_wordclouds(self):
+        # Generate wordclouds and save them as png images
+        start = time()
+        logging.info("Generating wordclouds")
+        if self.subchart_column != None:
+            for name, row in self.counts_df.iterrows():
+                print(name, row)
+                plt.close()
+                wc = self._get_wordcloud(row["count"])
+                fig = plt.figure(figsize=(38.4, 21.6), dpi=100)
+                fig.tight_layout()
+                plt.axis("off")
+                plt.imshow(wc, interpolation="bilinear")
+                path_fig = os.path.join(self.path, "wordcloud_" + name)
+                plt.savefig(path_fig, bbox_inches="tight", pad_inches=0, dpi=fig.dpi)
+            logging.info("Wordclouds generation done in {:.2f} seconds.".format(time() - start))
+        else:
+            plt.close()
+            wc = self._get_wordcloud(self.counts)
+            fig = plt.figure(figsize=(38.4, 21.6), dpi=100)
+            fig.tight_layout()
+            plt.axis("off")
+            plt.imshow(wc, interpolation="bilinear")
+            path_fig = os.path.join(self.path, "wordcloud")
+            plt.savefig(path_fig, bbox_inches="tight", pad_inches=0, dpi=fig.dpi)
```

Comment:
Same comment: The OS method do not work with UIF and remote folders. This will need to be rewritten to use the Dataiku API.

Reply:
The new version does solve the issue, but it uses an undocumented internal API ㊙️ .
I would rather use `from io import BytesIO` and then use the official `upload_data` method